### ISSUE TITLE
[C#] Implement new Concept Exercise: expression-bodied-members

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1145]
This issue describes how to implement the `expression-bodied-members` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student more advanced aspects of the Concept of Expression-bodied Members in C#.

## Learning objectives

- Know the difference between statements and expressions.
- Know how to define expression bodied members.

## Out of scope

- Performance difference between statements and expressions.

## Concepts

This Concepts Exercise's Concepts are:

- `expression-bodied-members`: know the difference between statements and expressions; know how to define expression bodied members.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `classes`: know how to define methods on classes.
- `constructors`: know how to define constructors.
- `properties`: know how to define properties.
- `indexers`: know how to define indexers.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

### Hints

- [Expressions][expressions]: explains what expressions are.
- [Expression-bodied members][expression-bodied-members]: what expression-bodied members are and how to define them.

### After

- [Expressions][expressions]: explains what expressions are.
- [Expression-bodied members][expression-bodied-members]: what expression-bodied members are and how to define them.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/expressions.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[expressions]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expressions
[expression-bodied-members]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members

**Comment by mikedamay** [comment]
@ErikSchierboom This a first go at a refactoring exercise.  You will recognize it.

Please have a look at [ExpressionBodiedMembers.cs](https://github.com/mikedamay/v3/blob/csharp/expression-bodied-members/languages/csharp/exercises/concept/expression-bodied-members/ExpressionBodiedMembers.cs) - the starting point, [`Example.cs`](https://github.com/mikedamay/v3/blob/csharp/expression-bodied-members/languages/csharp/exercises/concept/expression-bodied-members/.meta/Example.cs) - the refactored code, and [_insructions.md_](https://github.com/mikedamay/v3/blob/csharp/expression-bodied-members/languages/csharp/exercises/concept/expression-bodied-members/.docs/instructions.md).

**Comment by ErikSchierboom** [comment]
I do recognize it: `space-age`. What comments are you looking for? On this actual exercise or on the idea of the refactoring exercise?

**Comment by mikedamay** [comment]
Do you think this will work as an exercise?

**Comment by ErikSchierboom** [comment]
I'm not sure. There are some things I'm not sure about. My biggest concern is that this is an extremely trivial exercise, especially if one uses an IDE that does refactoring. Maybe paring it with another concept could help with that (not sure which concept that is). Maybe something expression related, like the ternary operator or `switch` expressions.

Regarding the actual expression-bodied members contents:

- I'd like to see properties that can be refactoring to an expression-bodied member
- The constructor can also use an expression-body in the example.
- It would be nice if we'd have a method that can't be converted to an expression-body (at least not in its current form)

**Comment by mikedamay** [comment]
I'll take that as a no.

I think ternary would go well.  I will add a property to something else and a failing use case.

Not so keen on constructors.  Side-effects don's seem to be in the spirit of expression-bodied-anything.  But, that could be just  me.

**Comment by mikedamay** [comment]
Thanks for the input

**Comment by ErikSchierboom** [comment]
> Not so keen on constructors. Side-effects don's seem to be in the spirit of expression-bodied-anything. But, that could be just me.

I understand this reasoning. I personally like using expression-bodies whenever I can, including `void` methods. But maybe that's also just me :) As long as we're mentioning that you _can_ use them in a constructor, I'm fine with not having them in the example.

P.S. Were you intending to use the `space-age` exercise? The idea for the Concept Exercises is for them to be newly designed, so that would rule out using a practice exercise (unless it is for inspiration).

**Comment by mikedamay** [comment]
closed by PR #1880 Expression-bodied-memebers
